### ISSUE TITLE
Property name in volume, config and secret are not prefixed with stack name

### ISSUE THREAD

**Issue by pszczekutowicz** [docker/cli#811]
**Description**
When property name is defined in objects like volumes, configs and secrets and there is no external property set to true objects name created by docker stack deploy are created without stack name prefix.

Lets say we have  project with several stacks named by domain and each stack contains several service. Some of services names are repeating. For example stacks **shop** and stack **warehouse** both of them have services named **orders**. Both stacks are located in separated directories and are deployed to same cluster.

stack **shop**
```
version: "3.5"
services:
  orders:
    image: orders:v1
    configs:
      - source: orders
        target: /usr/src/app/config/local.json
    volumes:
      - orders_cache:/usr/src/app/cache:rw
configs:
  orders:
    name: orders_v1
    file: ./orders.json
volumes:
  orders_cache:
    name: orders_cache_v1
```
stack **warehouse**
```
version: "3.5"
services:
  orders:
    image: orders:v1
    configs:
      - source: orders
        target: /usr/src/app/config/local.json
    volumes:
      - orders_cache:/usr/src/app/cache:rw
configs:
  orders:
    name: orders_v1
    file: ./orders.json
volumes:
  orders_cache:
    name: orders_cache_v1
```
As you can see in docker >=17.12 when name property is used there will be issue with configuration files. There can be similar issue with secrets. Above example also shows conflict with volumes.

Regarding volumes. For example I can imagine one stack deployed twice in one cluster with different names, lets say development and testing. If I create volume for database with property name as there will be no stack name prefix there will be one volume for two environments.

**Comment by dnephin** [comment]
> configs and secrets and there is no external property set to true objects name created by docker stack deploy are created without stack name prefix.

Yes, that's the entire purpose of having a `name` field. If you want scoped names, don't use the `name` field. 

I'm guessing that you are using the name field here because secrets and configs are immutable, and changing them requires that you change the name, so by using  the `name` field you only have to change one field instead of every reference. Is that correct?

You can use variables in the name field, which would let you scope them yourself, but then requires that you manage the values for those variables.

**Comment by pszczekutowicz** [comment]
You are right, documentation was misleading. In my opinion it was suggesting new business value with versioning in mind but it brings only interpolation to refactored externalization.

In above example my goal was to have single point of responsibility. If version of configuration was changed I needed to change only one value - name filed in config. With current solution I need to change config name as key in configs section and config name as source in service's configs.

Thank you for the tip regarding variables with stack name.
